### DISCUSSION TITLE
English support

### DISCUSSION THREAD

**Discussion by harshad1** [Kaljurand/K6nele#38] [Q&A] [answered]
Hi, 

I think this is a great app and will become important for anyone who wants a google/gapps free android experience. 

I would very much like English support and I am willing to donate a little towards this.

Thabks

**Comment by Kaljurand** [comment]
Thanks!

Support for English (or any other language) can be added/improved in various alternative ways.

For the speech keyboard:
- set up a server based on https://github.com/alumae/kaldi-gstreamer-server that offers English support. Change "Kõnele (fast recognition) / WebSocket URL" to point to the URL of your server. (In general, the URL management should be improved in Kõnele to support multiple URLs at the same time.)
- implement the `RecognitionService`-interface for an existing speech recognition server (or even offline service). This implementation would be a separate app, so in this case no changes to Kõnele are necessary: the service and its supported languages simply show up in the "Recognition languages & services" list allowing you to use them in the Kõnele speech keyboard.

For the voice search panel:
- set up a server based on https://github.com/alumae/ruby-pocketsphinx-server that offers English support. Change "Kõnele (grammar support) / HTTP URL" to point to the URL of your server.
- generalize Kõnele to allow any `RecognitionService` implementation to be used on the voice search panel (in the same was as currently possible on the speech keyboard).

Code contributions (via pull requests), concrete feature requests, etc. are welcome. (I personally do not take monetary donations for Kõnele work.)

**Comment by Kaljurand** [comment, marked as answer]
Starting with Kõnele v1.3 the instructions have become simpler.

Support for English (or any other language) can be added/improved in various alternative ways:
- implement the `RecognitionService`-interface for an existing speech recognition server (or even offline service). This implementation would be a separate app, i.e. no changes to Kõnele are necessary: the service and its supported languages simply show up in the "Recognition languages & services" list allowing you to use them in the Kõnele speech keyboard and voice search panel.
- set up a server based on https://github.com/alumae/kaldi-gstreamer-server that offers English support. Change "Kõnele (fast recognition) / WebSocket URL" to point to the URL of your server. (In general, the URL management should be improved in Kõnele to support multiple URLs at the same time.) See example in https://github.com/jcsilva/docker-kaldi-gstreamer-server/blob/22dcd22c6ed39962ebfcfc4182088054e075cc42/README.md
- set up a server based on https://github.com/alumae/ruby-pocketsphinx-server that offers English support. Change "Kõnele (grammar support) / HTTP URL" to point to the URL of your server.

When programmatically calling a Kõnele service then use the `ee.ioc.phon.android.extra.SERVER_URL` extra to override the server URL.

**Reply by Kaljurand** [reply]
An example of an app that implements the ``RecognizerService``-interface is https://github.com/Kaljurand/K6nele-service. You can use it as it is, and just override the "Server URL" to point to your server (that is based on https://github.com/alumae/kaldi-gstreamer-server ) and specify the supported locales in the "Server locales" field.

Or you can modify this app to wrap your own Android recognizer, whether offline or cloud-based.

**Comment by sydneyjd** [comment]
I think having English support would be great as well, and i'd host a server myself, but i don't have one to do it with. Has there been any headway at all with this? 
We do need more open alternatives to Google's voice search, which has gotten so popular.
Keep up the great work! :+1: 

-Sydney

**Comment by Kaljurand** [comment]
@sydneyjd I'm not planning to host a server, but if somebody is and wants to share a URL, I'll gladly include it in the Kõnele documentation. Even better would be if the server is wrapped by a `RecognitionService`-implementation that is made available as a separate app.

**Comment by khalidNash** [comment]
hi Nash here.......
i have been trying to implement konele as a part of my software application...i think its one of the best so far i have seen...but there are some issues still i am trying to solve regarding multi speaker recognition....and testing my app by including an audio file..plz share your views for the same.

**Comment by Kaljurand** [comment]
@khalidNash Please clarify what are you trying to achieve and what is missing in Kõnele, ideally as a separate issue.

**Comment by khalidNash** [comment]
Hi,

There are two things i would like to discuss first is regarding multi speaker recognition.....how can i use konele when many speakers are there and i would like to recognise them as speaker1,speaker2 and so on and seperately record there voice like in a a debate or conference?
now the second part is instead of directly recording the voice i would like to test the data by playing an audio file which it will convert to text....like we train the data as in case of kaldi we play the audio file of bill gates tedtalk...
i am in middle of a big project and konele has really served well but to make it more versatile we would have to include the multi speaker recognition and training the data by an audio file..
i would really appreciate any favours for this...

Thanks

**Comment by Kaljurand** [comment]
These should be separate issues, but here are some quick thoughts.

Multi speaker recognition. This would be an interesting feature, but currently there are no plans to implement this. Some of it should be done on the server side i.e. outside of the scope of K6nele. What is relevant for K6nele is the communication protocol with the server (e.g. each sent audio packet could be augmented with some meta data about the audio, and each received recognition result could contain a speaker ID if the server provides it), and the UI (i.e. how to display the information about multiple speakers).

Recording from the audio file. What is the exact use case? E.g. if you just want to test the server, then you could do it outside of K6nele with a simple Python script that streams the audio to the server. That said, it would be useful to generalize the code in 
https://github.com/Kaljurand/K6nele/tree/master/app/src/main/java/ee/ioc/phon/android/speak/service
to accept not just mic input but also audio files, i.e. replace calls to speechutils.AudioRecorder with some higher level interface (a la AudioProvider). This would also improve testing, but as an end-user feature I don't see a use case now.

If you are interested in these features and want to implement and offer them as pull requests, then please open separate issues.

**Comment by Kaljurand** [comment]
Closing as too vague.
